File: matlab_version/+algorithms/GeneticAlgorithm.m
classdef GeneticAlgorithm < algorithms.PlanningAlgorithm %
                                GENETICALGORITHM 遗传算法 MATLAB 实现

                                    properties PopulationSize CrossoverRate
                                        MutationRate ElitismCount end

                                            methods function obj =
    GeneticAlgorithm(config) % 构造函数 obj =
        obj @algorithms.PlanningAlgorithm(config);

obj.PopulationSize = 50;
obj.CrossoverRate = 0.8;
obj.MutationRate = 0.1;
obj.ElitismCount = 2;

if isfield (config, 'PopulationSize')
  , obj.PopulationSize = config.PopulationSize;
end if isfield (config, 'CrossoverRate'),
    obj.CrossoverRate = config.CrossoverRate;
end if isfield (config, 'MutationRate'), obj.MutationRate = config.MutationRate;
end end

    function bestSol =
        solve(obj, observations, satellites) % SOLVE 执行遗传算法 tic;

% 1. 预处理：建立 ID->Observation 映射以加速查询 obsMap = containers.Map();
            for
              i = 1 : length(observations) obsMap(observations(i).Id) =
                          observations(i);
            end obsIds = obsMap.keys;

            % 2. 初始化种群 %
                解表示：Assignments
                    Map(ObsID->SatID) population = cell(1, obj.PopulationSize);
            for
              i = 1 : obj.PopulationSize population{i} =
                  obj.generateRandomSolution(observations, satellites);
            end
            
            % 3. 迭代
            for gen = 1:obj.Config.MaxIterations
                if toc > obj.Config.TimeLimit
                    break;
            end

                % 评估 scores = zeros(1, obj.PopulationSize);
                for
                  i = 1 : obj.PopulationSize scores(i) =
                      obj.evaluate(population{i}, obsMap);

                % 更新全局最优 if isempty (obj.BestSolution) ||
                    scores(i) > obj.BestSolution.Objective obj.BestSolution =
                    population{i};
                obj.BestSolution.Objective = scores(i);
                end end

                    obj.History(end + 1) = obj.BestSolution.Objective;

                % 选择(锦标赛) newPop = cell(1, obj.PopulationSize);
                % 精英保留[~, sortedIdx] = sort(scores, 'descend');
                for
                  k = 1 : obj.ElitismCount newPop{k} = population{sortedIdx(k)};
                end

                    currentCount = obj.ElitismCount;
                while
                  currentCount < obj.PopulationSize parent1 =
                      obj.tournamentSelection(population, scores);
                parent2 = obj.tournamentSelection(population, scores);

                if rand ()
                  < obj.CrossoverRate[child1, child2] =
                      obj.crossover(parent1, parent2, obsIds);
                else
                  child1 = parent1;
                child2 = parent2;
                end

                    child1 = obj.mutate(child1, observations, satellites);
                child2 = obj.mutate(child2, observations, satellites);

                if currentCount
                  < obj.PopulationSize currentCount = currentCount + 1;
                newPop{currentCount} = child1;
                end if currentCount < obj.PopulationSize currentCount =
                    currentCount + 1;
                newPop{currentCount} = child2;
                end end

                    population = newPop;
                end

                    obj.Runtime = toc;
                bestSol = obj.BestSolution;
                end end

                    methods(Access = private) function sol =
                        generateRandomSolution(obj, observations, satellites)
                            sol.Assignments = containers.Map();
            for
              i = 1
                  : length(observations) if rand () >
                    0.5 %
                        假设 Observation 已包含可行卫星列表(Mock step needed) %
                        这里简化为随机分配给卫星 ID %
                        注意：Python 版本中 obs.satellite_id
                        是预计算好的单个可行解？
                        % 不，Python 中 obs 可能有多个窗口。 %
                        但 run_benchmark.py 中 BenchObservation 包含
                        'satellite_id' 属性，
                        %
                        这意味着 BenchObservation
                        本身就是一个(Task, Satellite,
                                     Time) 的元组，即一个可行窗口。
                        %
                        所以分配非常简单：如果要选这个
                        Observation(窗口)，它的 SatID 是固定的。

                        obs = observations(i);
            sol.Assignments(obs.Id) = obs.SatelliteId;
            end end sol.Objective = 0;
            % 待评估 end

                    function score = evaluate(obj, solution, obsMap) score = 0;
            keys = solution.Assignments.keys;
            for
              i = 1 : length(keys) if obsMap.isKey(keys{i}) obs =
                  obsMap(keys{i});
            score = score + obs.Score;
            end end end

                function selected = tournamentSelection(obj, pop, scores) k = 3;
            idx = randperm(length(pop), k);
            bestIdx = idx(1);
            for
              i = 2 : k if scores (idx(i)) > scores(bestIdx) bestIdx = idx(i);
            end end selected = pop{bestIdx};
            end

                function[c1, c2] =
                    crossover(obj, p1, p2, obsIds) % 单点交叉 point =
                        randi(length(obsIds));

            c1.Assignments = containers.Map();
            c2.Assignments = containers.Map();

            % 前半段
            for i = 1:point
                id = obsIds{i};
            if p1
              .Assignments.isKey(id), c1.Assignments(id) = p1.Assignments(id);
            end if p2.Assignments.isKey(id), c2.Assignments(id) =
                                                 p2.Assignments(id); end
            end
            
            % 后半段
            for i = (point+1):length(obsIds)
                id = obsIds{i};
            if p2
              .Assignments.isKey(id), c1.Assignments(id) = p2.Assignments(id);
            end if p1.Assignments.isKey(id), c2.Assignments(id) =
                                                 p1.Assignments(id);
            end end

                c1.Objective = 0;
            c2.Objective = 0;
            end

                function sol = mutate(obj, sol, observations, satellites) keys =
                    sol.Assignments.keys;
            % 随机移除
            for i = 1:length(keys)
                if rand() < obj.MutationRate
                    remove(sol.Assignments, keys{i});
                end
            end
            
            % 随机添加
            for i = 1:length(observations)
                obs = observations(i);
                if
                  ~sol.Assignments.isKey(obs.Id) &&
                      rand() < obj.MutationRate sol.Assignments(obs.Id) =
                      obs.SatelliteId;
                end end end end end

File: constellation_planning/algorithms/genetic_algorithm.py
# -*- coding: utf-8 -*-
"""遗传算法"""

import random
from typing import List, Any, Tuple
from .base import PlanningAlgorithm, Solution, AlgorithmConfig


class GeneticAlgorithm(PlanningAlgorithm):
    """遗传算法"""
    
    def __init__(
        self,
        config: AlgorithmConfig = None,
        population_size: int = 50,
        crossover_rate: float = 0.8,
        mutation_rate: float = 0.1,
        elitism_count: int = 2
    ):
        super().__init__(config)
        self.population_size = population_size
        self.crossover_rate = crossover_rate
        self.mutation_rate = mutation_rate
        self.elitism_count = elitism_count
    
    def get_name(self) -> str:
        return f"GA(pop={self.population_size}, cx={self.crossover_rate}, mut={self.mutation_rate})"
    
    def solve(
        self,
        observations: List[Any],
        satellites: List[Any],
        **kwargs
    ) -> Solution:
        self._start_timer()
        
        if self.config.random_seed:
            random.seed(self.config.random_seed)
        
        obs_list = list(observations)
        # 建立 ID 到 分数的映射
        self.score_map = {
            obs.id: getattr(obs, 'score', getattr(obs, 'priority', 1.0)) 
            for obs in obs_list
        }
        
        # 初始化种群
        population = [
            self._generate_random_solution(obs_list, satellites) 
            for _ in range(self.population_size)
        ]
        
        for generation in range(self.config.max_iterations):
            if self._is_time_exceeded():
                break
            
            # 评估
            for ind in population:
                self._update_best(ind)
            
            # 排序
            population.sort(key=lambda x: x.objective_value, reverse=True)
            
            # 精英保留
            new_population = population[:self.elitism_count]
            
            # 生成新个体
            while len(new_population) < self.population_size:
                # 选择
                parent1 = self._tournament_selection(population)
                parent2 = self._tournament_selection(population)
                
                # 交叉
                if random.random() < self.crossover_rate:
                    child1, child2 = self._crossover(parent1, parent2, obs_list)
                else:
                    child1, child2 = parent1.copy(), parent2.copy()
                
                # 变异
                self._mutate(child1, obs_list, satellites)
                self._mutate(child2, obs_list, satellites)
                
                new_population.extend([child1, child2])
            
            population = new_population[:self.population_size]
            if self.best_solution:
                self.history.append(self.best_solution.objective_value)
        
        return self.best_solution
    
    def _calculate_score(self, solution: Solution) -> float:
        """计算解的总分"""
        return sum(self.score_map.get(obs_id, 1.0) for obs_id in solution.assignments)

    def _generate_random_solution(
        self,
        observations: List[Any],
        satellites: List[Any]
    ) -> Solution:
        """生成随机解"""
        solution = Solution()
        for obs in observations:
            if random.random() > 0.5 and hasattr(obs, 'satellite_id'):
                solution.assignments[obs.id] = obs.satellite_id
        solution.objective_value = self._calculate_score(solution)
        return solution
    
    def _tournament_selection(
        self,
        population: List[Solution],
        tournament_size: int = 3
    ) -> Solution:
        """锦标赛选择"""
        candidates = random.sample(population, min(tournament_size, len(population)))
        return max(candidates, key=lambda x: x.objective_value)
    
    def _crossover(
        self,
        parent1: Solution,
        parent2: Solution,
        observations: List[Any]
    ) -> Tuple[Solution, Solution]:
        """单点交叉"""
        child1 = Solution()
        child2 = Solution()
        
        crossover_point = random.randint(0, len(observations))
        
        for i, obs in enumerate(observations):
            if i < crossover_point:
                if obs.id in parent1.assignments:
                    child1.assignments[obs.id] = parent1.assignments[obs.id]
                if obs.id in parent2.assignments:
                    child2.assignments[obs.id] = parent2.assignments[obs.id]
            else:
                if obs.id in parent2.assignments:
                    child1.assignments[obs.id] = parent2.assignments[obs.id]
                if obs.id in parent1.assignments:
                    child2.assignments[obs.id] = parent1.assignments[obs.id]
        
        child1.objective_value = self._calculate_score(child1)
        child2.objective_value = self._calculate_score(child2)
        return child1, child2
    
    def _mutate(
        self,
        solution: Solution,
        observations: List[Any],
        satellites: List[Any]
    ) -> None:
        """变异"""
        for obs in observations:
            if random.random() < self.mutation_rate:
                if obs.id in solution.assignments:
                    del solution.assignments[obs.id]
                else:
                    if hasattr(obs, 'satellite_id'):
                        solution.assignments[obs.id] = obs.satellite_id
        
        solution.objective_value = self._calculate_score(solution)
